### PR TITLE
Added CMake option OGS_EIGEN_DYNAMIC for using dynamically allocated eigen matrices

### PR DESCRIPTION
**Pull request by bilke**
See #987.

- [x] Update Jenkins jobs to use fixed-size matrices

### CODE REVIEW

**Comment by bilke** [comment]
Fails as assumed in debug with dynamic eigen matrices.

In general to speed things up in Travis I suggest to test with only one compiler which reduces parallel builds to 3. Testing of the other compiler can then be done on Jenkins.

**Comment by bilke** [comment]
Tried to exclude specific build combinations without success. Excluding seems not to work together with global environment variables...

**Comment by chleh** [comment]
Maybe it's only a matter of taste, but I'd prefer a very verbose name, like `OGS_EIGEN_DYNAMIC_SHAPE_MATRICES`.

File: NumLib/Fem/ShapeMatrixPolicy.h [hunk 1]
```diff
@@ -107,8 +107,11 @@ struct EigenDynamicShapeMatrixPolicy
 
 /// Default choice of the ShapeMatrixPolicy.
 template <typename ShapeFunction, unsigned GlobalDim>
+#if EIGEN_DYNAMIC_MATRICES
```
Comment (chleh) [inline NumLib/Fem/ShapeMatrixPolicy.h]: Could you pull the `template<...>` line inside the `#ifdef', please. I think, that's more clear.

**Comment by chleh** [comment]
I suggest that pull requests be built with dynamic shape matrices only, since that compiles faster than fixed-size matrices and thereby reduces load on the CI systems. Then, when PRs get merged, there could be builds with fixed-size matrices, too, in order to catch all errors.
What do you think?

**Comment by chleh** [comment]
@bilke when it works in general, I'm fine with it :fast_forward:

**Comment by bilke** [comment]
Any volunteers to fix the crashing eigen test before merging this?

**Comment by chleh** [comment]
Maybe the test really should be fixed before merging this PR. It seems to be the only one that crashes. I can have a look at it by end of this week.

**Comment by chleh** [comment]
@bilke: By your last commit, will there now be any test of fixed-size eigen matrices? I think they have to be tested, too, because: (1) They might use entirely different template functions inside our OGS code, (2) there will be compile-time checks that matrix dimensions in producty match only for them.
In general I agree that we should default to dynamic matrices in order to save much time, but there should be a check of fixed-size matrices, e.g., after a PR has been merged.

**Comment by bilke** [comment]
Yes I will update Jenkins jobs to test fixed-size matrices.

**Pull request merged**